pkg/server/node.go

```diff
@@ -858,10 +868,82 @@ func (n *Node) addStore(ctx context.Context, store *kvserver.Store) {
 		// initialization process.
 		log.Fatal(ctx, "attempting to add a store without a version")
 	}
+	store.TODOEngine().RegisterDiskSlowCallback(func(info pebble.DiskSlowInfo) {
+		n.onStoreDiskSlow(ctx, store.StoreID(), info)
+	})
 	n.stores.AddStore(store)
 	n.recorder.AddStore(store)
 }
 
+func (n *Node) onStoreDiskSlow(
+	ctx context.Context, storeID roachpb.StoreID, info pebble.DiskSlowInfo,
+) {
+	// Emit a disk slow event for each slow store that reports a disk slow event.
+	// After diskSlowClearInterval since the last disk slow event, we emit a disk
+	// slowness cleared event. Any other disk slow events for the same store in
+	// between are ignored as duplicates.
+	//
+	// TODO(bilal): Store stats on events observed and coalesced events.
+	const diskSlowClearInterval = 30 * time.Second
+	if storeID == 0 {
+		// Uninitialized store; cannot coalesce events for an unknown store.
+		return
+	}
+	n.diskSlowCoalescerMu.Lock()
+	last, ok := n.diskSlowCoalescerMu.lastDiskSlow[storeID]
+	n.diskSlowCoalescerMu.lastDiskSlow[storeID] = timeutil.Now()
+	n.diskSlowCoalescerMu.Unlock()
+	// NB: A zero / nonexistent value for last means that we are not currently running
+	// the below async task, and this is a new disk slowness event that needs a
+	// start/clear pair emitted for it.
+	if ok && last != (time.Time{}) {
+		// This is a duplicate event for this store.
+		return
+	}
+
+	ev := &eventpb.DiskSlownessDetected{}
+	ev.StoreID = int32(storeID)
+	ev.NodeID = int32(n.Descriptor.NodeID)
+	ev.CommonDetails().Timestamp = timeutil.Now().UnixNano()
+
+	n.logStructuredEvent(ctx, logpb.EventPayload(ev))
+
+	_ = n.stopper.RunAsyncTask(ctx, "clear-disk-slow", func(ctx context.Context) {
+		ticker := time.NewTicker(1 * time.Second)
+		defer ticker.Stop()
+		for {
+			select {
+			case <-ticker.C:
+				clearing := false
+				n.diskSlowCoalescerMu.Lock()
+				last := n.diskSlowCoalescerMu.lastDiskSlow[storeID]
+				if timeutil.Since(last) < diskSlowClearInterval {
+					clearing = true
+					delete(n.diskSlowCoalescerMu.lastDiskSlow, storeID)
+				}
+				n.diskSlowCoalescerMu.Unlock()
+
+				if !clearing {
+					continue
+				}
+
+				// Clear the event and break out of the loop.
+				ev := &eventpb.DiskSlownessCleared{}
+				ev.StoreID = int32(storeID)
+				ev.NodeID = int32(n.Descriptor.NodeID)
+				ev.CommonDetails().Timestamp = timeutil.Now().UnixNano()
+
+				n.logStructuredEvent(ctx, logpb.EventPayload(ev))
```

Comment:
Is this blocking, or best effort? What happens if we're attempting to log slowness for the store that is the LH for the event log, and its disk is indefinitely stalled?

Reply:
This method does a log to the node's logger (where a stall would have a different flushing goroutine wait instead), and it does an INSERT asynchronously in a different goroutine into the eventlog table. So it seems like we should be covered here, and shouldn't unintentionally hold up the disk health checker.

Reply:
On closer inspection the `log.StructuredEvent` could under some cases write directly to the logger file sink, even though the SQL write will always be async. So to decouple this from the handler coming all the way from Pebble, I moved the entire `logStructuredEvent` call into the async task below, and left a comment about it.